File: regions/system/library/dart/eventstore/lib/src/snapshot.dart
class Snapshot {
  final String streamId;
  final int version;
  final Object? state;
  final DateTime createdAt;

  const Snapshot({
    required this.streamId,
    required this.version,
    this.state,
    required this.createdAt,
  });
}

abstract class SnapshotStore {
  Future<void> saveSnapshot(Snapshot snapshot);
  Future<Snapshot?> loadSnapshot(String streamId);
}

class InMemorySnapshotStore implements SnapshotStore {
  final Map<String, Snapshot> _snapshots = {};

  @override
  Future<void> saveSnapshot(Snapshot snapshot) async {
    _snapshots[snapshot.streamId] = snapshot;
  }

  @override
  Future<Snapshot?> loadSnapshot(String streamId) async {
    return _snapshots[streamId];
  }
}

File: regions/system/library/rust/eventstore/src/snapshot.rs
use async_trait::async_trait;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::{EventStoreError, StreamId};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Snapshot {
    pub stream_id: String,
    pub version: u64,
    pub state: serde_json::Value,
    pub created_at: DateTime<Utc>,
}

#[async_trait]
pub trait SnapshotStore: Send + Sync {
    async fn save_snapshot(&self, snapshot: Snapshot) -> Result<(), EventStoreError>;
    async fn load_snapshot(
        &self,
        stream_id: &StreamId,
    ) -> Result<Option<Snapshot>, EventStoreError>;
}
